File: PopulatingRightPointer.cpp
/**
Intuition:
#####################################################################
Solution 1: Do level order traversal, in each level, assign the curr popped node to the next element in the queue.
Time Complexity : O(N), N = Number of nodes in the tree
Space Complexity : O(N/2), N/2 = Number of leaf nodes.
#####################################################################
Solution 2: Keep assing the next pointer using the left and right properties of the nodes.
Time Complexity : O(N), N = Number of nodes in the tree
Space Complexity : O(1), H = Height of the tree
#####################################################################
*/

//Solution 1
class Solution {
public:
    Node* connect(Node* root) {
        queue <Node*> queue;
        if ( root == NULL) return root;
        queue.push(root);
        
        Node* curr = NULL;
        Node* nextPointer = NULL;
        while ( queue.size()!= 0){
            int size = queue.size();
            for ( int i =0; i < size; i++){
                
                curr = queue.front();
                queue.pop();
                
                if ( i == size-1){
                    curr->next = NULL;
                }
                else{
                    nextPointer = queue.front();
                    curr->next = nextPointer;
                }
                
                if (curr->left != NULL) queue.push(curr->left);
                if (curr->right != NULL) queue.push(curr->right);
                
            }
        }
        return root;
    
    }
};

//Solution 2
class Solution {
public:
    Node* connect(Node* root) {
        if ( root == NULL) return NULL;
        Node* level = root;
        Node* curr;
        while ( level->left != NULL ){
            curr = level;
            while ( curr != NULL){

                curr->left->next = curr->right;
                if ( curr->next != NULL)
                    curr->right->next = curr->next->left;

                curr = curr->next;
            }
            level = level->left;
        }
        return root;
    }
};

File: PopulatingRightPointer.java
class Solution {
    public Node connect(Node root) {
        Queue<Node> q= new LinkedList<>();
         if (root == null) {
            return root;
        }
        q.add(root);
        int size;
        while(!q.isEmpty()){
            size= q.size();
            Node curr;
            for(int i=0; i< size; i++){
                curr= q.poll();
                if (i < size - 1) {
                    curr.next = q.peek();
                }
                if (curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            }
        }
        return root;
    }
}

//time - O(n)
//space - O(n)

//another approach

class Solution {
 
    public Node connect(Node root) {
        Node level_start=root;
        while(level_start!=null){
            Node cur=level_start;
            while(cur!=null){
                if(cur.left!=null) cur.left.next=cur.right;
                if(cur.right!=null && cur.next!=null) cur.right.next=cur.next.left;
                
                cur=cur.next;
            }
            level_start=level_start.left;
        }
        return root;
    }  
} 
//time -O(n)
//space- O(1)